runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -2029,7 +2037,7 @@ public Iterable<Entry<K, V>> read() {
                   .removeIf(
                       entry ->
                           entry.getValue().existence == KeyExistence.KNOWN_NONEXISTENT
-                              && !localRemovals.contains(entry.getKey()));
+                              && !entry.getValue().removedLocally);
               return Iterables.unmodifiableIterable(mergedCachedEntries());
             } else {
               MultimapIterables<K, V> local = mergedCachedEntries();
```

Comment:
I'm worried about memory usage in this case. If entries is not weighted it means it is a paginating iterable to windmill IIUC. But the for-each above on 2003 (in new version) is going to read the full iterable and add all the keys to the keyStateMap.  It seems like a new iterator should be returned that wraps the existing iterable so that it can lazily do that as iteration occurs

Reply:
Ah that makes a lot of sense. Updated the code so that if entries is not weighted, it will not be evaluated immediately.